### PR TITLE
feat: add multiple getters mode in `generate_getter`

### PR DESCRIPTION
**Pull request by feniljain**
This commit adds two modes to generate_getter action.
First, the plain old working on single fields.
Second, working on a selected range of fields.

Should partially solve #13246 
If this gets approved will create a separate PR for setters version of the same

### Points to help in review:

- `generate_getter_from_record_info` contains code which is mostly taken from assist before refactor
- Same goes for `parse_record_fields`
- There are changes in other assists, as one of the methods in utils named `find_struct_impl` is changed, before it used to accept a single `fn_name`, now it takes a list of function names to check against. All old impls are updated to create a small list to pass their single element.

### Assumptions:

- If any of the fields have an implementation, the action will quit.

### CODE REVIEW

**Comment by feniljain** [comment]
Hey @Veykril , would love to get a review on this!

File: crates/ide-assists/src/handlers/generate_delegate_methods.rs [hunk 1]
```diff
@@ -51,14 +51,14 @@ pub(crate) fn generate_delegate_methods(acc: &mut Assists, ctx: &AssistContext<'
         Some(field) => {
             let field_name = field.name()?;
             let field_ty = field.ty()?;
-            (format!("{field_name}"), field_ty, field.syntax().text_range())
+            (format!("{}", field_name), field_ty, field.syntax().text_range())
```
Comment (Veykril) [inline crates/ide-assists/src/handlers/generate_delegate_methods.rs:54]: `field_name.to_string()`

File: crates/ide-assists/src/handlers/generate_delegate_methods.rs [hunk 2]
```diff
@@ -51,14 +51,14 @@ pub(crate) fn generate_delegate_methods(acc: &mut Assists, ctx: &AssistContext<'
...
         }
         None => {
             let field = ctx.find_node_at_offset::<ast::TupleField>()?;
             let field_list = ctx.find_node_at_offset::<ast::TupleFieldList>()?;
             let field_list_index = field_list.fields().position(|it| it == field)?;
             let field_ty = field.ty()?;
-            (format!("{field_list_index}"), field_ty, field.syntax().text_range())
+            (format!("{}", field_list_index), field_ty, field.syntax().text_range())
```
Comment (Veykril) [inline crates/ide-assists/src/handlers/generate_delegate_methods.rs:61]: `field_list_index.to_string()`

File: crates/ide-assists/src/handlers/generate_delegate_methods.rs [hunk 3]
```diff
@@ -153,12 +151,12 @@ pub(crate) fn generate_delegate_methods(acc: &mut Assists, ctx: &AssistContext<'
                             Some(cap) => {
                                 let offset = strukt.syntax().text_range().end();
                                 let snippet = render_snippet(cap, impl_def.syntax(), cursor);
-                                let snippet = format!("\n\n{snippet}");
+                                let snippet = format!("\n\n{}", snippet);
```
Comment (Veykril) [inline crates/ide-assists/src/handlers/generate_delegate_methods.rs:154]: Unnecessary change

File: crates/ide-assists/src/handlers/generate_getter.rs [hunk 4]
```diff
@@ -72,88 +75,263 @@ pub(crate) fn generate_getter_mut(acc: &mut Assists, ctx: &AssistContext<'_>) ->
...
+
+        let node = match node {
+            syntax::NodeOrToken::Node(n) => n,
+            syntax::NodeOrToken::Token(t) => t.parent()?,
+        };
+
+        let mut parent_struct: Option<ast::Struct> = None;
+        if let Some(struct_node) = node.ancestors().find(|it| ast::Struct::can_cast(it.kind())) {
+            parent_struct = ast::Struct::cast(struct_node);
+        }
```
Comment (Veykril) [inline crates/ide-assists/src/handlers/generate_getter.rs:116]:
```suggestion
        let parent_struct = node.ancestors().find_map(ast::Struct::cast)?;
```

File: crates/ide-assists/src/handlers/generate_getter.rs [hunk 5]
```diff
@@ -72,88 +75,263 @@ pub(crate) fn generate_getter_mut(acc: &mut Assists, ctx: &AssistContext<'_>) ->
...
+
+        // If we don't have a parent struct, something is wrong, OPERATION ABORT 🚨
+        if parent_struct.is_none() {
+            return None;
+        }
+
+        let (info_of_record_fields, field_names) =
+            extract_and_parse_record_fields(&node, ctx.selection_trimmed(), mutable)?;
+
+        (parent_struct.unwrap(), info_of_record_fields, field_names)
```
Comment (Veykril) [inline crates/ide-assists/src/handlers/generate_getter.rs:126]: Given we can just `?` as in my earlier review, the unwrap becomes unnecessary (as well as the explicit `is_none()` check on line 119

File: crates/ide-assists/src/handlers/generate_getter.rs [hunk 6]
```diff
@@ -72,88 +75,263 @@ pub(crate) fn generate_getter_mut(acc: &mut Assists, ctx: &AssistContext<'_>) ->
...
+
+    // No record fields to do work on :(
+    if info_of_record_fields.len() == 0 {
+        return None;
     }
-    let impl_def = find_struct_impl(ctx, &ast::Adt::Struct(strukt.clone()), fn_name.as_str())?;
+
+    // `find_struct_impl` looks for impl of a struct, but this also has additional feature
+    // where it takes a list of function names and check if they exist inside impl_, if
+    // even one match is found, it returns None
```
Comment (Veykril) [inline crates/ide-assists/src/handlers/generate_getter.rs:146]: I feel like this comment should be a doc comment on `find_struct_impl` instead :)

File: crates/ide-assists/src/handlers/generate_getter.rs [hunk 7]
```diff
@@ -72,88 +75,263 @@ pub(crate) fn generate_getter_mut(acc: &mut Assists, ctx: &AssistContext<'_>) ->
...
     };
-    let target = field.syntax().text_range();
+
+    let mut target: TextRange = info_of_record_fields[0].target;
+
+    // Computing collective text range of all record fields in selected region
+    info_of_record_fields
+        .clone()
+        .iter()
+        .for_each(|record_field_info| target = target.cover(record_field_info.target));
```
Comment (Veykril) [inline crates/ide-assists/src/handlers/generate_getter.rs:161]:
```suggestion
    let target: TextRange = info_of_record_fields
        .iter()
        .map(|record_field_info| record_field_info.target)
        .reduce(|acc, target| acc.cover(target));
```

File: crates/ide-assists/src/handlers/generate_getter.rs [hunk 8]
```diff
@@ -72,88 +75,263 @@ pub(crate) fn generate_getter_mut(acc: &mut Assists, ctx: &AssistContext<'_>) ->
...
         target,
         |builder| {
+            let record_fields_count = info_of_record_fields.len();
+
             let mut buf = String::with_capacity(512);
 
-            if impl_def.is_some() {
-                buf.push('\n');
+            // Check if an impl exists
+            if let Some(ref impl_def) = getter_info.impl_def {
```
Comment (Veykril) [inline crates/ide-assists/src/handlers/generate_getter.rs:176]:
```suggestion
            if let Some(impl_def) = &getter_info.impl_def {
```

File: crates/ide-assists/src/utils.rs [hunk 9]
```diff
@@ -334,7 +334,7 @@ fn calc_depth(pat: &ast::Pat, depth: usize) -> usize {
 pub(crate) fn find_struct_impl(
     ctx: &AssistContext<'_>,
     adt: &ast::Adt,
-    name: &str,
+    names: Vec<String>,
```
Comment (Veykril) [inline crates/ide-assists/src/utils.rs:337]:
```suggestion
    names: &[String],
```
We don't need ownership here, so `&[...]` is more general

File: crates/ide-assists/src/utils.rs [hunk 10]
```diff
@@ -362,20 +362,20 @@ pub(crate) fn find_struct_impl(
...
+        if has_any_fn(impl_blk, names) {
             return None;
         }
     }
 
     Some(block)
 }
 
-fn has_fn(imp: &ast::Impl, rhs_name: &str) -> bool {
+fn has_any_fn(imp: &ast::Impl, names: Vec<String>) -> bool {
```
Comment (Veykril) [inline crates/ide-assists/src/utils.rs:373]:
```suggestion
fn has_any_fn(imp: &ast::Impl, names: &[String]) -> bool {
```

File: crates/ide-assists/src/utils.rs [hunk 11]
```diff
@@ -362,20 +362,20 @@ pub(crate) fn find_struct_impl(
...
     if let Some(il) = imp.assoc_item_list() {
         for item in il.assoc_items() {
             if let ast::AssocItem::Fn(f) = item {
                 if let Some(name) = f.name() {
-                    if name.text().eq_ignore_ascii_case(rhs_name) {
+                    if names.contains(&name.text().to_string().to_ascii_lowercase()) {
```
Comment (Veykril) [inline crates/ide-assists/src/utils.rs:378]:
```suggestion
                    if names.iter().any(|n| n.eq_ignore_ascii_case(name.text())) {
```
This is a lot cheaper as we save on the `to_string` and `to_ascii_lowercase` allocations

File: crates/ide-assists/src/utils.rs [hunk 12]
```diff
@@ -334,7 +334,7 @@ fn calc_depth(pat: &ast::Pat, depth: usize) -> usize {
 pub(crate) fn find_struct_impl(
```
Comment (Veykril) [inline crates/ide-assists/src/utils.rs:334]: One possible improvement (no need to do this now) would be to have this return/tell which names are already occupied and which aren't. That way we could still show the assist and fill the missing ones.

File: crates/ide-assists/src/handlers/generate_getter.rs [hunk 13]
```diff
@@ -72,88 +75,263 @@ pub(crate) fn generate_getter_mut(acc: &mut Assists, ctx: &AssistContext<'_>) ->
...
+                // Check if impl is empty
+                if let Some(assoc_item_list) = impl_def.assoc_item_list() {
+                    if assoc_item_list.assoc_items().next().is_some() {
+                        // If not empty then only insert a new line
+                        buf.push('\n');
+                    }
+                }
+            }
+
+            for i in 0..record_fields_count {
```
Comment (Veykril) [inline crates/ide-assists/src/handlers/generate_getter.rs:186]:
```suggestion
            for (i, record_field_info) in info_of_record_fields.iter().enumerate() {
```
this is more idiomatic as we no longer need to idnex into `info_of_record_fields`

File: crates/ide-assists/src/handlers/generate_getter.rs [hunk 14]
```diff
@@ -72,88 +75,263 @@ pub(crate) fn generate_getter_mut(acc: &mut Assists, ctx: &AssistContext<'_>) ->
...
+            parent_struct = ast::Struct::cast(struct_node);
+        }
+
+        // If we don't have a parent struct, something is wrong, OPERATION ABORT 🚨
+        if parent_struct.is_none() {
+            return None;
+        }
+
+        let (info_of_record_fields, field_names) =
+            extract_and_parse_record_fields(&node, ctx.selection_trimmed(), mutable)?;
```
Comment (Veykril) [inline crates/ide-assists/src/handlers/generate_getter.rs:124]:
```suggestion
            extract_and_parse_record_fields(&parent_struct, ctx.selection_trimmed(), mutable)?;
```
No need to use the untyped node here

File: crates/ide-assists/src/handlers/generate_getter.rs [hunk 15]
```diff
@@ -72,88 +75,263 @@ pub(crate) fn generate_getter_mut(acc: &mut Assists, ctx: &AssistContext<'_>) ->
...
+        info.mutable.then(|| "mut ").unwrap_or_default(),
+        ty,
+        body,
+    );
+
+    buf
+}
+
+fn extract_and_parse_record_fields(
+    node: &SyntaxNode,
```
Comment (Veykril) [inline crates/ide-assists/src/handlers/generate_getter.rs:296]:
```suggestion
    node: &ast::Struct,
```
Then we can use the typed API for ast::Struct to fetch the fields instead of using `.children(..)`

File: crates/ide-assists/src/handlers/generate_delegate_methods.rs (hunk 3 above)
Comment (feniljain) [reply crates/ide-assists/src/handlers/generate_delegate_methods.rs:154]: I made this change cause I got a conflict from another PR, which did the same thing all over different assists

File: crates/ide-assists/src/handlers/generate_getter.rs (hunk 4 above)
Comment (feniljain) [reply crates/ide-assists/src/handlers/generate_getter.rs:116]: Nice catch!

File: crates/ide-assists/src/handlers/generate_getter.rs (hunk 6 above)
Comment (feniljain) [reply crates/ide-assists/src/handlers/generate_getter.rs:146]: Makes sense, it would be better there

File: crates/ide-assists/src/handlers/generate_getter.rs (hunk 7 above)
Comment (feniljain) [reply crates/ide-assists/src/handlers/generate_getter.rs:161]: Missed a `?` at the end, also nice, I can use a map to convert into new iterator and then reduce upon it, ah I really need to get a hang of using iterators properly 😅

File: crates/ide-assists/src/utils.rs (hunk 11 above)
Comment (feniljain) [reply crates/ide-assists/src/utils.rs:378]: Nice optimization!

File: crates/ide-assists/src/utils.rs (hunk 12 above)
Comment (feniljain) [reply crates/ide-assists/src/utils.rs:334]: That's a nice idea, I can pick this up in a different PR

File: crates/ide-assists/src/handlers/generate_getter.rs (hunk 13 above)
Comment (feniljain) [reply crates/ide-assists/src/handlers/generate_getter.rs:186]: Nice observation!

File: crates/ide-assists/src/utils.rs (hunk 9 above)
Comment (feniljain) [reply crates/ide-assists/src/utils.rs]: Nice catch!

**Comment by feniljain** [comment]
Thanks for in depth and amazing review, have resolved all of them now!

**Comment by Veykril** [comment]
Thanks!
Small note, try to not force push when you address reviews, unless you need to due to a rebase or similar. A PR can have multiple commits if need be. (This makes it easier to see what you have changed in regards to a review). 

@bors r+

**Comment by feniljain** [comment]
>  Small note, try to not force push when you address reviews, unless you need to due to a rebase or similar. A PR can have multiple commits if need be. (This makes it easier to see what you have changed in regards to a review).

Oh will keep that in mind, thnx for the wisdom :bow:
